tests/integration/test_bundle.py

```diff
@@ -137,3 +203,144 @@ async def test_multiple_kratos_external_idp_integrators(ops_test: OpsTest):
 
     action_output = await get_redirect_uri_action.wait()
     assert "redirect-uri" in action_output.results
+
+
+async def test_kratos_scale_up(ops_test: OpsTest) -> None:
+    """Check that kratos works after it is scaled up."""
+    app_name = "kratos"
+
+    app = ops_test.model.applications[app_name]
+
+    await app.scale(3)
+
+    await ops_test.model.wait_for_idle(
+        raise_on_blocked=False,
+        raise_on_error=False,
+        status="active",
+        timeout=2000,
+    )
+
+    admin_address = await get_app_address(ops_test, TRAEFIK_ADMIN_APP)
+    health_check_url = (
+        f"https://{admin_address}/{ops_test.model.name}-{app_name}/admin/health/ready"
+    )
+    resp = requests.get(health_check_url, verify=False)
+
+    assert resp.status_code == 200
+
+
+async def test_hydra_scale_up(ops_test: OpsTest) -> None:
+    """Check that hydra works after it is scaled up."""
+    app_name = "hydra"
+
+    app = ops_test.model.applications[app_name]
+
+    await app.scale(3)
+
+    await ops_test.model.wait_for_idle(
+        raise_on_blocked=False,
+        raise_on_error=False,
+        status="active",
+        timeout=2000,
+    )
+
+    admin_address = await get_app_address(ops_test, TRAEFIK_ADMIN_APP)
+    health_check_url = f"https://{admin_address}/{ops_test.model.name}-{app_name}/health/ready"
+    resp = requests.get(health_check_url, verify=False)
+
+    assert resp.status_code == 200
+
+
+async def test_create_hydra_client(ops_test: OpsTest, state: State, dex: str) -> None:
+    """Register a client on hydra."""
+    # This is a hack, we just need a server to be running on the redirect_uri
+    # so that when we get redirected there we don't get an connection_refused
+    # error.
+    redirect_uri = join(dex, "some", "path")
+    app = ops_test.model.applications["hydra"]
+    action = await app.units[0].run_action(
+        "create-oauth-client",
+        **{
+            "redirect-uris": [redirect_uri],
+            "grant-types": ["authorization_code", "client_credentials"],
+        },
+    )
+
+    res = (await action.wait()).results
+
+    assert res["client-id"]
+    assert res["client-secret"]
+    assert res["redirect-uris"] == redirect_uri
+
+    state.client_id = res["client-id"]
+    state.client_secret = res["client-secret"]
+    state.redirect_uri = res["redirect-uris"]
+
+
+async def test_authorization_code_flow(
+    ops_test: OpsTest,
+    state: State,
+    dex: str,
+    dex_user_email: str,
+    dex_user_password: str,
+    page: Page,
+) -> None:
+    hydra_url = await get_reverse_proxy_app_url(ops_test, TRAEFIK_PUBLIC_APP, "hydra")
```

Comment:
Nitpick:

- Declare the application names in a grouped way? Like the `TRAEFIK_PUBLIC_APP` in the global scope at the top, or group all applications in an `enum` in case we will have more applications coming to the bundle in the future?

Reply:
Makes sense, I added a namedtuple.